### ISSUE TITLE
SetUpFixture not respecting NonParallelizable tag on TestFixtures.

### ISSUE THREAD

**Issue by woodylaurence** [nunit/nunit#2454]
It seems since version 3.8.0, the inclusion of a SetUpFixture causes issues when trying to run multiple fixtures marked as parallelizable but some fixtures marked as non-parallelizable.

I have parallelism set at the assembly level with the following in my AssemblyInfo.cs:
`[assembly: Parallelizable(ParallelScope.Fixtures)]`

Here's my test setup:
```
namespace Tests
{
	[SetUpFixture]
	public class TestsSetUpFixture
	{

	}

	[TestFixture]
	public class TestFixture1
	{
		[Test]
		public void TestFixture1_Test()
		{
			Thread.Sleep(1000);
		}
	}

	[TestFixture]
	[NonParallelizable]
	public class TestFixture2
	{
		[Test]
		public void TestFixture2_Test()
		{
			Thread.Sleep(1000);
		}
	}

	[TestFixture]
	public class TestFixture3
	{
		[Test]
		public void TestFixture3_Test()
		{
			Thread.Sleep(1000);
		}
	}
}
```

All three test fixtures get run in parallel, I would expect 1 and 3 to run in parallel whilst 2 is run separately. This is the behaviour when running under version 3.7.1, however upgrading to 3.8 causes them all to run in parallel (the same is also true under 3.8.1).

We get the same problem if you remove the parallelism setting from AssemblyInfo and instead mark TestFixture1 and TestFixture3 as [Parallelizable(ParallelScope.Self)].

As work around, for now,  you can change the ParallelScope in AssemblyInfo to Children instead of Fixtures and the behaviour returns to what is expected where all TestFixture 2 waits for the other 2 fixtures to finish executing. If you have more than 1 test in a fixture  you do end up with tests within fixtures running in parallel; this can be further fixed by adding the ParallelScope.Self above the fixtures which need to run in parallel.

**Comment by CharliePoole** [comment]
You are correct in your expectations. In this simple situation, I would expect fixtures 1 and 3 to run to completion in parallel, followed by fixture 2 executing in isolation.

This appears to duplicate at least part of #2395 but I'm keeping it alive since you have a nice simple repro case.

**Comment by CharliePoole** [comment]
@woodylaurence Just to be clear... if you remove the SetUpFixture, it works as expected?

**Comment by CharliePoole** [comment]
If the SetUpFixture is causing the problem, please try these two alternatives:

1. Remove the assembly-level `ParallelizableAttribute` and put it on the SetUpFixture.
2. Leave the assembly-level attribute but move the SetUpFixture into the global namespace - i.e. outside of the namespace braces.

If either or both of these fixes the problem, it will give us more information. I recently fixed a similar problem with global-level SetUpFixtures. It could be that I missed a case.

**Comment by woodylaurence** [comment]
Thanks, I will confirm the behaviour of those two suggestions when I'm back at work on Monday.

**Comment by woodylaurence** [comment]
I have tried both removing the assembly-level attribute and putting it on SetUpFixture as well as leaving the assembly-level attribute and moving the SetUpFixture into the global namespace; neither of these fixed the issue, the behaviour remained the same.

Thanks.

**Comment by CharliePoole** [comment]
OK, thanks for checking. I'll set up a test bed and see what I can come up with. We don't have good tests for parallel execution. It calls for something a bit more than a unit test of course.

**Comment by CharliePoole** [comment]
I reproduced the problem.

**Comment by CharliePoole** [comment]
@nunit/framework-team This is a simple issue with big repercussions. It highlights many problems that are showing up at the same moment in other issues around parallelism. I'd appreciate your reading what I've found and giving me your comments...

1. Start and end shift messages in the log do not appear in the right order when a SetUpFixture is present. Sometimes shifts are nested or overlap. Shifts, of course, are supposed to be self-contained with no two shifts active at the same time.

2. We end shifts when there may still be items running, which may directly execute additional child items on the same thread.

3. When we end a shift, we immediately restore all queues. But there may be items to run in the other queues, which do not pertain to that shift.

Here's how I think we got here. We had a pretty clean object model for parallel execution in "shifts" to start out. I refactored it after the initial implementation to make it into a hierarchy: dispatcher => shifts => queues and workers.

Subsequently however, in fixing some problems, we bolted on some things that don't quite match the object model. I did some of it myself, in fact. In particular, we wanted to have a stack of real queues, so that each non-parallel item with parallel children could finish all it's work before anyone else got to run. Logically, that called for a new level in the hierarchy to control one set of queues. As a matter of expediency, we didn't do that and now we are suffering.

I'm working two ideas in parallel: trying to think of what the model ought to have been while patching further the existing model. Ideas are welcome.

**Comment by ChrisMaddock** [comment]
2 and 3 match up to what I discovered digging in to this a couple of weeks ago. I also agree the model needs adapting - saving and restoring queues in a stack-esque fashion feeds strangely meta - but I'm not totally sure what the best solution is either!

**Comment by CharliePoole** [comment]
I put the queues into a stack inside `WorkItemQueue` so that I would not have to create new workers. The same workers keep pulling from the `WorkItemQueue` but the underlying implementation`changes. Writing it down like that makes the fragility pretty obvious.

I can imagine two alternate models...

1. Dispatcher => SetOfShifts => Shifts => Queues => Workers
2. MasterController => Dispatchers => Shifts => Queues => Workers

In the first the `SetOfShifts` encapsulates the isolation. It could have some other name, of course. In the second a master controller creates an entirely new dispatcher when we need to have isolation of a fixture.

All this creating of new objects could get inefficient. We can make it less so if we ensure that all the classes are cheap to create and that any resource usage like thread creation is postponed till needed. We could also do some predictive analysis and use isolation when it's not needed. For example, a non-parallel fixture without any parallel or STA tests in it doesn't need to enqueue anything - it can just execute all the items directly. My inclination would be to solve the problems and live with any inefficiency until we see how bad it is, then optimize.

**Comment by rprouse** [comment]
> My inclination would be to solve the problems and live with any inefficiency until we see how bad it is, then optimize.

I agree with that as long as we are postponing thread creation until needed wherever possible. I assume you meant that when you said "ensure that all the classes are cheap to create and that any resource usage like thread creation is postponed till needed."

Like @ChrisMaddock I also saw 2 and 3 while testing his possible fix. As for selecting from the two models that you present above, I think it requires too much knowledge of the problem space, but the second `MasterController` seems to be a simpler approach as it will be a central point of coordination, but maybe I am reading too much into the names 😄

**Comment by CharliePoole** [comment]
I seem to have found a simple fix. It should go in for now, I suppose. In order to find it, I did a lot of boilerplate stuff: more logging, changes to test utilities, etc. I'm going to do a PR with only the resulting change that fixes the bug. I'll follow that with a second that does the boilerplate, non-functional stuff and then look at the overall structure. I have hopes my fix will fix some other issues as well.
